[logging] Log runtime autotuning timing to scuba

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141919

See test plan in internal diff [D66679369](https://our.internmc.facebook.com/intern/diff/D66679369)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

masnesral committed:
[logging] Log cudagraphify and autotuning timing to scuba

[ghstack-poisoned]

masnesral commented:
@masnesral has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66679369).

masnesral committed:
Update on "[logging] Log cudagraphify and autotuning timing to scuba"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D66679369](https://our.internmc.facebook.com/intern/diff/D66679369)

[ghstack-poisoned]

masnesral commented:
@masnesral has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66679369).

masnesral committed:
Update on "[logging] Log cudagraphify and autotuning timing to scuba"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D66679369](https://our.internmc.facebook.com/intern/diff/D66679369)

[ghstack-poisoned]

masnesral commented:
@masnesral has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66679369).

masnesral commented on torch/_inductor/codegen/triton.py:
```diff
@@ -2918,6 +2918,7 @@ def inductor_meta_common():
             "min_split_scan_rblock": config.triton.min_split_scan_rblock,
             "spill_threshold": config.triton.spill_threshold,
             "store_cubin": config.triton.store_cubin,
+            "compile_id": str(torch._guards.CompileContext.current_compile_id()),
```
This is necessary if we want the compile_id to show up in the tlparse / perfetto metadata. I assume we do.

masnesral commented on torch/_inductor/compile_fx.py:
```diff
@@ -1158,10 +1159,7 @@ def cudagraphify(
-                "cudagraphify",
-                log_pt2_compile_event=True,
-            ), dynamo_utils.preserve_rng_state():
+            with dynamo_utils.preserve_rng_state():
```
Pretty sure the placement of this dynamo_timed was wrong. It's confusing because of the closure. I moved it to the cudagraphify call in cudagraph_trees.py. The result looks better to my eyes; see the perfetto trace linked in the test plan for the internal diff D66679369. cc @BoyuanFeng

masnesral commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -416,25 +422,41 @@ def cudagraphify(
-    )
+    # Even though this is technically runtime, we want to capture these overheads for
+    # "compilation metrics" logging.
+    metrics_context = get_metrics_context()
```
The introduction of this metrics_context is what gives us the dynamo_compile scuba log that encompasses the "cudagraphify" and autotuning timings.

masnesral replied:
<img width="695" alt="Screenshot 2024-12-03 at 5 24 59 PM" src="https://github.com/user-attachments/assets/99626b92-31b7-485c-bf59-f411b3e975b0">

Well what I'm seeing experimentally is something like this. i.e., I see the "CUDAGraphTreeManager.run_eager" timer that you added, but no "CUDAGraphTreeManager.record_function". And neither seem to account for the full time of the cudagraphify

masnesral commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -880,7 +880,11 @@ def clone_args(self, *args, **kwargs) -> Tuple[List[Any], Dict[str, Any]]:
+            log_pt2_compile_event=True,
+            metadata={"kernel_name": self.inductor_meta.get("kernel_name", None)},
+            dynamo_compile_column_us="runtime_triton_autotune_time_us",
+            compile_id=self.inductor_meta.get("compile_id", None),
```
The compile_id here is the other half of a comment I made above. I save away the compile_id in the inductor_meta as part of the codegen, then retrieve it here and pass to dynamo_timed.

jamesjwu commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -416,25 +422,41 @@ def cudagraphify(
+    metrics_context = get_metrics_context()
+    with metrics_context, dynamo_timed(
+        "cudagraphify",
+        log_pt2_compile_event=True,
```
This should only happen O(1) times per compile, right? Just confirming for retention reasons in PT2 Compile Events.

masnesral replied:
That's right

jamesjwu reviewed (approved):
See internal comments

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -334,6 +339,9 @@ def _foo(...):
     - dynamo_compile_column_us: If provided, updates the specified CompilationMetrics
       field to be logged to dyname_compile column. We expect all columns to be _us;
       therefore, the field name must end with "_us".
+    - compile_id: In the typical case, this parameter should not be needed. Use to
```
It's kind of annoying but I'm not sure I would prefer the logical alternative, which is to force you to make a compile context and make it the current one before entering dynamo_timed. The problem is that this is not cheap and the logic here needs to happen in the fastpath execution.

ezyang reviewed (changes requested):
I think the autotuning logic generates a dynamo_compile entry per autotune, which isn't acceptable

masnesral commented:
> I think the autotuning logic generates a dynamo_compile entry per autotune, which isn't acceptable

@ezyang do you see that in the links I provided on D66679369 or are you just expecting that given the implementation here? I'm not seeing that experimentally at least. For example, on the random benchmark I used in the test plan (dcgan), there's autotuning that happens for one of the compile_id's in the forward and backward. So for that given compile_id, we get these entries:

1) An entry for the dynamo compile for the forward.
2) An entry for the autotuning of the forward
3) An entry for the entire compile of the backward
4) An entry for the autotuning of the backward

For each autotuning step, there are several kernels tuned and those are all aggregated into the single log.

masnesral commented:
@ezyang I need your input on another aspect: there's a tlparse test failing with `--strict` because the name passed to          `torch._logging.trace_structured()` is "compilation_metrics", and it seems like we expect some extra fields to be available (for example `compliant_custom_ops`). For these entries that originate from the autotuning, I'm not trying to populate those fields in the tlparse log. How do you suggest to change things? Should these runtime logs use a different name?

ezyang commented:
Re `--strict`, it seems like tlparse cannot actually deal with None compliant_custom_ops

```
2024-12-03T06:41:18.1174274Z Parser compilation_metrics failed: Encountered rendering error on line 60, column 19. Reason: Expected an array for path 'm.compliant_custom_ops' but found a non-iterable value.
```

we do need to fix this, but it is better to fix it in tlparse. Update the template code to handle the case whe compliant_custom_ops is None and then rev the tlparse version in this diff.

ezyang commented:
> https://github.com/pytorch/pytorch/pull/141919#issuecomment-2515413724

I have reread through the trace and now I have a different problem: I think I want a version of the trace where you do NOT have cudagraphs enabled but you DO have autotuning enabled. I suspect you will get lots of dynamo_compile entries in that case.

masnesral commented:
> I suspect you will get lots of dynamo_compile entries in that case.

@ezyang, yeah there's a problem. It's not that we get lots of dynamo_compile entries; it's that there's no outer context in that case, so there's no log created at all. Hmmm, then am I correct that there's no natural place to insert the contextmanager that will a) sit above all the autotuning, but b) emits a log only once? For example, I can put the logging contextmanager around `runtime_wrappers.forward`, but that would create a bunch of empty logs.

I dunno, one way I can think to fix is to update the MetricsContext utility to emit a log only in the case that at least some field has been populated.

ezyang commented:
> Hmmm, then am I correct that there's no natural place to insert the contextmanager that will a) sit above all the autotuning, but b) emits a log only once?

Yes, this is the problem we discussed a month ago. I thought our tentative agreement was to use the same strategy as @aorenste's Triton autotune cache bundler to have a naked event extrude out from the compilation which gets closed in the finally block at the end of the actual runtime code.

masnesral commented:
> Yes, this is the problem we discussed a month ago.

@ezyang Ok, sorry I must've missed the agreement discussion or it went over my head. I did ping @aorenste since he's mentioned in related tasks and messenger threads and we discussed how the AutotuneCacheBundler couldn't be used off the shelf. But now I think you mean: follow the pattern and not reuse directly. 

So looking again at the AutotuneCacheBundler, I think we have 2 options on the table. They look similar in spirit to me:
1) Do something like the AutotuneCacheBundler: Mark down when we first start compiling, accumulate info during autotuning, hook into the point after the callable executes. On that last step: if info has accumulated, log it.
2) What I vaguely proposed above: Put the MetricsContext contextmanager around the forward and backward, accumulate info during autotuning, log on exit from the contextmanager (but only if info has accumulated).

I'd lean towards 2 because it works the same as what we have now with only a minor modification; we don't have two mechanisms for this logging. Do you not like it? Do you think it won't work?

masnesral commented:
Or maybe I put the contextmanager here: https://github.com/pytorch/pytorch/blob/d24b1475204419dd2dc25161ec17198b0e04ced3/torch/_inductor/output_code.py#L466
But same idea: log only if there have been some dynamo_timed calls that have given us something to log

ezyang commented:
The main problem with putting it in the runtime wrapper (which I agree it would work) is you're adding a fixed cost to EVERY iteration of the compiled code, even when no autotuning occurs. An @aorenste strategy still has a tax, but only a branch in the finally block of the compiled code. At least he isn't creating an entire new context manager object and blah blah blah.

masnesral committed:
Update on "[logging] Log runtime autotuning timing to scuba"


See test plan in internal diff D66679369

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D66679369](https://our.internmc.facebook.com/intern/diff/D66679369)

[ghstack-poisoned]

masnesral commented:
@masnesral has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66679369).

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -1065,10 +1097,16 @@ def safe_str(item: Any) -> str:
+    if compilation_metrics.is_forward is False:
+        name = "bwd_" + name
+    if compilation_metrics.is_runtime is True:
+        name = name + "_runtime"
```
@ezyang so the new names would be "compilation_metrics_runtime" and "bwd_compilation_metrics_runtime". But I don't know what I need to change to get those to show up in manifold, e.g., https://fburl.com/z3ydyrdm

ezyang replied:
You'll need to update tlparse, I think

masnesral committed:
Update on "[logging] Log runtime autotuning timing to scuba"


See test plan in internal diff D66679369

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D66679369](https://our.internmc.facebook.com/intern/diff/D66679369)

[ghstack-poisoned]

masnesral committed:
Update on "[logging] Log runtime autotuning timing to scuba"


See test plan in internal diff D66679369

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D66679369](https://our.internmc.facebook.com/intern/diff/D66679369)

[ghstack-poisoned]

masnesral committed:
Update on "[logging] Log runtime autotuning timing to scuba"


See test plan in internal diff D66679369

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D66679369](https://our.internmc.facebook.com/intern/diff/D66679369)

[ghstack-poisoned]

masnesral commented:
@masnesral has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66679369).

masnesral committed:
Update on "[logging] Log runtime autotuning timing to scuba"


See test plan in internal diff D66679369

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D66679369](https://our.internmc.facebook.com/intern/diff/D66679369)

[ghstack-poisoned]

masnesral commented:
@masnesral has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66679369).

masnesral committed:
Update on "[logging] Log runtime autotuning timing to scuba"


See test plan in internal diff D66679369

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D66679369](https://our.internmc.facebook.com/intern/diff/D66679369)

[ghstack-poisoned]

masnesral commented:
@masnesral has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66679369).

ezyang commented on torch/_guards.py:
```diff
@@ -64,6 +64,17 @@ class CompileId(NamedTuple):
+        Factory method that creates a CompileId from its string representation.
+        """
+        try:
+            frame_id, frame_compile_id = compile_id.split("/")
```
cc @xmfan 

@masnesral I'm pretty confused why you need to parse string...

masnesral replied:
@ezyang 

Only because the compile_id makes its way to the benchmarking method as a string via the inductor_meta, e.g.,

```
@triton_heuristics.pointwise(
    size_hints={'x': 2097152},
    filename=__file__,
    triton_meta={'signature': {'in_ptr0': '*fp16', 'out_ptr0': '*fp32', 'xnumel': 'i32'}, 'device': DeviceProperties(type='cuda', index=0, multi_processor_count=108, cc=80, major=8, regs_per_multiprocessor=65536, max_threads_per_multi_processo
r=2048, warp_size=32), 'constants': {}, 'configs': [AttrsDescriptor(divisible_by_16=(0, 1, 2), equal_to_1=())]},
    inductor_meta={'autotune_hints': set(), 'kernel_name': 'triton_poi_fused__to_copy_12', 'mutated_arg_names': [], 'optimize_mem': True, 'no_x_dim': False, 'num_load': 1, 'num_reduction': 0, 'backend_hash': '406CE938696801C1383AAC67BD6068B4A6
9521593A9E5D0F4469315E0D6B6E77', 'are_deterministic_algorithms_enabled': False, 'assert_indirect_indexing': True, 'autotune_local_cache': True, 'autotune_pointwise': True, 'autotune_remote_cache': None, 'force_disable_caches': False, 'dynamic_
scale_rblock': True, 'max_autotune': False, 'max_autotune_pointwise': False, 'min_split_scan_rblock': 256, 'spill_threshold': 16, 'store_cubin': False, 'is_forward': False, 'compile_id': '2/0'},
    min_elem_per_thread=0
)
@triton.jit
def triton_poi_fused__to_copy_12(in_ptr0, out_ptr0, xnumel, XBLOCK : tl.constexpr):
...
```

I can probably pass it as a CompileId type instead, then make sure we import from `torch._guards` in the generated code. Or rather, the generated code currently has only a few "lightweight" imports, mostly from `torch._inductor.runtime`, so probably the definition of CompileId would want to move somewhere else. wdyt?

xmfan replied:
Can we get rid of TraceId and have a single mutable object CompileId? Could we centralize the unpacking logic as a method in CompileId? I am landing https://github.com/pytorch/pytorch/pull/141907 which will change it to a dataclass

ezyang replied:
Consolidating might be fine but I am 👎  on making it mutable. Useful to have an immutable type with value semantics that I don't have to worry about being mutated under me.

ezyang commented on torch/_dynamo/metrics_context.py:
```diff
@@ -126,3 +126,42 @@ def add_to_set(self, metric: str, value: Any) -> None:
+
+
+class RuntimeMetricsContext:
+    def __init__(self, on_exit: OnExitType):
```
I'm not a big fan of this indirection haha, mostly cause now I have to dig to find where on_exit is defined

masnesral replied:
Established in `MetricsContext`, the thinking being that MetricsContext` is (currently) generalized and completely agnostic of how it's used to ultimately log. Which also made the unit tests actual unit tests and easy to write without any mocking. Happy to change both if it bothers you.

masnesral committed:
Update on "[logging] Log runtime autotuning timing to scuba"


See test plan in internal diff D66679369

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D66679369](https://our.internmc.facebook.com/intern/diff/D66679369)

[ghstack-poisoned]

masnesral committed:
Update on "[logging] Log runtime autotuning timing to scuba"


See test plan in internal diff D66679369

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D66679369](https://our.internmc.facebook.com/intern/diff/D66679369)

[ghstack-poisoned]

xmfan commented on torch/_inductor/codegen/triton.py:
```diff
@@ -3037,6 +3038,8 @@ def inductor_meta_common():
             "min_split_scan_rblock": config.triton.min_split_scan_rblock,
             "spill_threshold": config.triton.spill_threshold,
             "store_cubin": config.triton.store_cubin,
+            "compile_id": str(compile_id) if compile_id else None,
```
is there some interaction with caching that we need to worry about here?

masnesral replied:
@xmfan Yes, maaaybe. That's a question I had too. This will create logs using the compile ID available during codegen. So if we get a cache hit, but in a context with a different CompileId, then it will be quite confusing to see the runtime logs associated with a CompileId from some other run. I don't know how to fix it...

eellison replied:
I just ran into bad cache hits because of this. This means we'll never get a cache hit based on source code for this. and if we do get cache hit, as you said, it will be confusingly labeled. is there another approach that we could do here ?

eellison replied:
concretely, we rely on caching of source code for the benchmarking of epilogue fusion. i am seeing with `BlenderbotSmallForConditionalGeneration` cache misses where we should see cache hits.

masnesral replied:
Dang sorry. I didn't think through that it could actually break caching. This change is all about internal logging and there's some discussion happening here: https://fburl.com/workplace/ga37uahr

masnesral replied:
@eellison, oh interesting. I think it's easy enough to grab the compile_id at the time we load the Python module and then save it as an attribute. What I don't immediately see is how to look it up, e.g., at the point of timing the autotuning. Any tips?

eellison replied:
Pass it in as an arg to `async_compile.wait(globals())` maybe.

masnesral committed:
Update on "[logging] Log runtime autotuning timing to scuba"


See test plan in internal diff D66679369

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D66679369](https://our.internmc.facebook.com/intern/diff/D66679369)

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge

masnesral committed:
Update on "[logging] Log runtime autotuning timing to scuba"


See test plan in internal diff [D66679369](https://our.internmc.facebook.com/intern/diff/D66679369)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge
